rust/ql/lib/codeql/rust/elements/internal/TypeParamImpl.qll

```diff
@@ -28,11 +28,16 @@ module Impl {
     /** Gets the position of this type parameter. */
     int getPosition() { this = any(GenericParamList l).getTypeParam(result) }
 
-    private int nrOfDirectTypeBounds() {
-      result = this.getTypeBoundList().getNumberOfBounds()
-      or
-      not this.hasTypeBoundList() and
-      result = 0
+    private TypeBound getTypeBoundAt(int i, int j) {
+      exists(TypeBoundList tbl | result = tbl.getBound(j) |
+        tbl = this.getTypeBoundList() and i = 0
+        or
+        exists(WherePred wp |
+          wp = this.(TypeParamItemNode).getAWherePred() and
+          tbl = wp.getTypeBoundList() and
+          wp = any(WhereClause wc).getPredicate(i)
+        )
```

Comment:
I feel like this case should simplify to `tbl = this.(TypeParamItemNode).getWherePred(i).getTypeBoundList()` (except that `getWherePred(i)` doesn't exist there).  I don't feel strongly if this is a pain to change.

Reply:
I'll give it a try. I had that predicate before, but ran into non-monotonic recursion. But it should be doable as long as the `getWherePred(i)` is not actually used in path resolution itself.

Reply:
I agree that what you propose is a lot cleaner, but the `getWherePred` on `TypeParamItemNode` would have to use `rank` and then we got non-monotonic recursion 😢